### ISSUE TITLE
Support for optional with default

### ISSUE THREAD

**Issue by malaverdiere** [scallop/scallop#70]
It would be nice to have an option type that has the following behavior:
- no --opt = opt.isSupplied is false opt() gives None
- --opt = opt.isSupplied is true, opt() gives Some(default)
- --cg abc = opt.isSupplied is true, opt() gives Some(abc)

I am copy-pasting the WIP from titanpad

```
def optWithDefault[A](default: A)(implicit conv: ValueConverter[A], tt: TypeTag[Option[A]]) = 
 new ValueConverter[Option[A]] {
    def parse(s: List[(String, List[String])]) = {
      s match {
        case Nil => Right(Some(None))
        case (_, Nil) :: Nil => Right(Some(Some(default)))
        case (_, v :: Nil) :: Nil => conv.parse(v).right.map(_.map(Some))
      }
   }
   val tag = tt
   val argType = ArgType.LIST
 }
usage:
// definitely need a shorter name for this converter
val cg = opt[Option[String]]()(optWithDefault("default.file"))
cg() // should return an option, that is None if there was no option call, Some(default) if there was an empty option call, Some(arg) otherwise.
```

**Comment by Rogach** [comment]
What will be the best name for this converter, in your opinion?

**Comment by malaverdiere** [comment]
DefaultOpt?

It would be better to have this I think: `cg = defaultOpt[String](default="default.file,...)`

**Comment by Rogach** [comment]
No, introducing `.defaultOpt` will only add clutter and confusion. I think that this syntax will be better:

``` scala
val cg = opt[Option[String]]()(optDefault("default.file"))
```

It's not exactly obvious, but the use case isn't exactly frequent, either.

**Comment by Rogach** [comment]
I implemented a feature, but tweaked the semantics a little (so that `.isSupplied` returns sane results):
- no `--opt` = opt.isSupplied is false, opt.get gives None
- `--opt` = opt.isSupplied is true, opt.get gives Some(default)
- `--opt abc` = opt.isSupplied is true, opt.get gives Some(abc)

I published a snapshot (for scala 2.10.x):

``` scala
resolvers += "sonatype snapshots" at "https://oss.sonatype.org/content/repositories/snapshots"
libraryDependencies += "org.rogach" %% "scallop" % "0.9.3~2~SNAPSHOT"
```

You can look at test cases [here](https://github.com/Rogach/scallop/commit/ee194c2d54bd85ddd7962b9e58a7e96e78cb3cd2#L1R57).

**Comment by Rogach** [comment]
I'm closing the issue and merging the change now. If you will have any additional thoughts about it, please let me know!

**Comment by malaverdiere** [comment]
I tried it and it looks fine. The only thing missing is that the default value is not displayed when I do --help

**Comment by Rogach** [comment]
Yes, `--help` stuff slipped under my radar. I'll try to fix it (but it'll probably take some time, maybe I'll do it next week).

**Comment by Rogach** [comment]
@malaverdiere - I thought about it a bit more, and one thing troubles me - how should output in help look like?

If I format it the same way the default value for options is formatted now, user may be confused - for example, he may think that if he completely omits the option, he will get the listed default value, while it isn't the case.

What do you think about it?

**Comment by malaverdiere** [comment]
Good point, which I didn't think about.

Maybe something like:

```
--option [setting]                 this is what the option is       default setting is XYZ if --option is specified without another setting
```

**Comment by Rogach** [comment]
Hm. I think that it's way too wordy, and it would be much simpler if developer added this text manually (this way he may choose the exact wording, suitable for the situation). Do you agree?

**Comment by malaverdiere** [comment]
That's what I've done personally. I am a big fan of automation though :)

**Comment by Rogach** [comment]
Well, all good programmers are huge fans of automation :) But in this case, automation may actually get in the way - so I feel it is unnecessary.

I'm closing the issue now. If you will have any new ideas about it, feel free to reopen!
